### PR TITLE
fix(gradle): Fix compatibility with react-native-legal

### PR DESCRIPTION
**Pull request by romtsn**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :scroll: Description

- This moves the entire script logic from `afterEvaluate` (which is too late to register tasks) into `androidComponents.onVariants`. Since our tasks only run when there's an android app being build, it's a good place to hook our logic into. It also gives us direct access to the variant, so we don't have to search for it and build a `releases` map anymore. The tasks are configured per-variant, omitting the `debug` variant.
- Register our tasks lazily to avoid early resolution via `register` instead of `create`

I tested it locally both `debug` and `release` builds and the `react-native-legal` json file is correctly generated in both cases. Also, our tasks run correctly for the release build, you can see here I uploaded sourcemaps: https://sentry-sdks.sentry.io/settings/projects/sentry-react-native/source-maps/23317267-0d24-5f07-9852-28630b8cf98a/

The only thing that probably needs more testing is modules collection, because it generated an empty json locally for me, but perhaps I was doing something unexpected, because I was running gradle builds directly as opposed calling `yarn/npx`. If someone could verify it works, that'd be great.

## :bulb: Motivation and Context

Closes https://github.com/getsentry/sentry-react-native/issues/5236

## :green_heart: How did you test it?
manually

## :pencil: Checklist

- [ ] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [ ] I updated the docs if needed.
- [ ] I updated the wizard if needed.
- [ ] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: packages/core/sentry.gradle [hunk 1]
```diff
@@ -22,11 +22,265 @@ interface InjectedExecOps {
 
 def config = project.hasProperty("sentryCli") ? project.sentryCli : [];
 
+plugins.withId('com.android.application') {
+    def ac = extensions.getByName("androidComponents")
+
+    ac.onVariants(androidComponents.selector().all()) { v ->
```
Comment (lucas-zimerman) [inline packages/core/sentry.gradle]: Q: v is the release?

**Comment by lucas-zimerman** [comment]
Thank you for the PR! is it ready for review?

**Comment by romtsn** [comment]
@lucas-zimerman not yet, I will mark it as ready soonish, still sorting out some stuff

**Comment by romtsn** [comment]
@lucas-zimerman this is ready for review now.

File: packages/core/sentry.gradle (hunk 1 above)
Comment (romtsn) [reply packages/core/sentry.gradle]: v is `variant`, but since we have one down there already, I just named it `v`
